File: 10-Recursion-and-Backtracking/java/03-SudokuSolver.java
class Solution {

    int[] rows = new int[9];
    int[] cols = new int[9];
    int[] boxes = new int[9];

    public void solveSudoku(char[][] board) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') continue;

                int digit = board[i][j] - '1';
                int bit = 1 << digit;

                rows[i] |= bit;
                cols[j] |= bit;
                boxes[(i / 3) * 3 + (j / 3)] |= bit;
            }
        }

        dfs(board);
    }

    boolean dfs(char[][] board) {
        int row = -1;
        int col = -1;
        int minChoices = 10;

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.') continue;

                int box = (i / 3) * 3 + (j / 3);
                int available = ~(rows[i] | cols[j] | boxes[box]) & 511;
                int choices = Integer.bitCount(available);

                if (choices < minChoices) {
                    minChoices = choices;
                    row = i;
                    col = j;
                }
            }
        }

        if (row == -1) return true;

        int box = (row / 3) * 3 + (col / 3);
        int available = ~(rows[row] | cols[col] | boxes[box]) & 511;

        while (available != 0) {
            int bit = available & -available;
            int digit = Integer.numberOfTrailingZeros(bit);

            board[row][col] = (char) (digit + '1');

            rows[row] |= bit;
            cols[col] |= bit;
            boxes[box] |= bit;

            if (dfs(board)) return true;

            rows[row] ^= bit;
            cols[col] ^= bit;
            boxes[box] ^= bit;

            board[row][col] = '.';

            available &= (available - 1);
        }

        return false;
    }
}

File: 10-Recursion-and-Backtracking/python/03_sudoku_solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        rows = [0] * 9
        cols = [0] * 9
        boxes = [0] * 9

        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    continue
                digit = int(board[i][j]) - 1
                bit = 1 << digit
                rows[i] |= bit
                cols[j] |= bit
                boxes[(i // 3) * 3 + (j // 3)] |= bit

        def dfs():
            row, col = -1, -1
            min_choices = 10

            for i in range(9):
                for j in range(9):
                    if board[i][j] != '.':
                        continue
                    box = (i // 3) * 3 + (j // 3)
                    available = ~(rows[i] | cols[j] | boxes[box]) & 511
                    choices = available.bit_count()
                    if choices < min_choices:
                        min_choices = choices
                        row, col = i, j

            if row == -1:
                return True

            box = (row // 3) * 3 + (col // 3)
            available = ~(rows[row] | cols[col] | boxes[box]) & 511

            while available:
                bit = available & -available
                digit = (bit - 1).bit_length()

                board[row][col] = str(digit + 1)
                rows[row] |= bit
                cols[col] |= bit
                boxes[box] |= bit

                if dfs():
                    return True

                rows[row] ^= bit
                cols[col] ^= bit
                boxes[box] ^= bit
                board[row][col] = '.'

                available &= available - 1

            return False

        dfs()
